File: Trees/LinkedList2018F.cc
#include <iostream>
using namespace std;

class LinkedList {
private:
	class Node {
	public:
		int v;
		Node* next;
		Node(int v, Node* n) : v(v), next(n) {}
	};
	Node* head;
public:
  LinkedList() : head(nullptr) {}
	~LinkedList() {
		for (Node* p = head; p != nullptr; ) {
			Node* q = p;
			p = p->next;
			delete q;
		}
	}
  LinkedList(const LinkedList& orig) {
		if (orig.head == nullptr) {
			head = nullptr;
			return;
		}
		head = new Node(orig.head->v, nullptr);
		Node*q = head;
		for (Node* p = orig.head->next; p != nullptr; p = p->next, q = q->next) {
      q->next = new Node(p->v, nullptr);
		}
	}

	LinkedList& operator =(const LinkedList& orig) {
		LinkedList copy(orig); // calls copy constructor
		swap(head, copy.head);
		return *this;
	}

  void addEnd(int v) {
		if (head == nullptr) {
			head = new Node(v, nullptr);
			return;
		}
		Node* p;
		for(p = head; p->next != nullptr; p= p->next)
			;
		p->next = new Node(v, nullptr);
	}

	void addStart(int v) {
		head = new Node(v, head);
	}
	void removeEnd() {
		if (head == nullptr)
			return;
		if (head->next == nullptr) {
			delete head;
			head = nullptr;
			return;
		}
		Node* p;
		Node* q;
		for (p = head, q = p->next; q->next != nullptr; p = q, q = q ->next)
			;
		delete q;
		p->next = nullptr;
	}
	void removeStart(){
		Node* temp = head;
		head = head->next;
		delete temp;
	}
	friend ostream& operator <<(ostream& s, const LinkedList& list) {
		for (Node* p = list.head; p != nullptr; p = p->next)
			s << p->v;
		return s;
	}
};

int main() {
	// a = 9876543210
	LinkedList a;
	for (int i = 0; i < 10; i++)
		a.addStart(i);
	cout << a;
	LinkedList b = a; // 9876543210 copy constructor
	b.addStart(19);
	cout << a;
	LinkedList c(a); // 9876543210copy constructor
	LinkedList d;
	d.addStart(5);
	d = c; // operator =
	c = c;
}
	

File: LinkedLists/LinkedList2018F.java
public class LinkedList2018F {
	private static class Node { // LinkedList.Node
		int v;
		Node next;
	}
	private Node head;
	public LinkedList2018F() {
		head = null;
	}

	public void addEnd(int v) {
		if (head == null) {
			addStart(v);
			return;
		}
		// invariant: head is not null, list has at least 1 element
		Node p;
		for (p = head; p.next != null; p= p.next)
			;
		Node temp = new Node();
	  temp.v = v;
		temp.next = null;
		p.next = temp;		
	}
	public void addStart(int v) {
    Node temp = new Node();
		temp.v = v;
		temp.next = head;
		head = temp;
	}
	public void removeEnd() {
		if (head == null)
			return;
		if (head.next == null) {
			head = null;
			return;
		}
		// invariant: there are at least 2 elements
		Node p, q;
		for (p = head, q = p.next; q.next != null; p = q, q = q.next)
			;
		p.next = null;
	}
	public void removeStart() {
		head = head.next;
	}
	public String toString() {
		StringBuilder b = new StringBuilder();
		for (Node p = head; p != null; p = p.next)
			b.append(p.v).append(' ');
		return b.toString();
	}
	public static void main(String[]args) {
		LinkedList2018F a = new LinkedList2018F();
		for (int i = 10; i > 0; i--)
			a.addStart(i);
		System.out.println(a);
	}
}

